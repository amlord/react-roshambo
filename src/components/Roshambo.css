/* Roshambo */
.Roshambo {
  background: #d5d5d5;
  font-family: EncodeSansNarrowLight, helvetica, arial, sans-serif;
}

/* Roshambo Panel */
.Roshambo__container {
  max-width: 800px;
  padding: 20px;
  margin: 0 auto;
}

.Roshambo__panel {
  padding: 20px;
  background: #fff;
  border-radius: 4px;
  text-align: center;
  box-shadow: 0 3px 6px 1px rgba(0,0,0,0.14);
}

.Roshambo__title {
  margin: 0;
  line-height: 1em;
}

.Roshambo__subtitle {
  font-size: 0.8rem;
  font-weight: normal;
  margin-bottom: 20px;
  margin-top: 5px;
}

.Roshambo__game {
  font-size: 0;
}

/* Roshambo Footer */
.Roshambo__footer {
  color: #333;
  text-align: center;
  margin-top: 10px;
  font-weight: 100;
}