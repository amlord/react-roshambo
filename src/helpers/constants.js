export const ROCK = 1;
export const PAPER = 2;
export const SCISSORS = 3;

export const WIN = 'win';
export const DRAW = 'draw';
export const LOSS = 'loss';
