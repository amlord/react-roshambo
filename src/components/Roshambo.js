import React, { Component } from 'react';
import rnd  from 'pure-random';

import logo from '../logo.svg';
import './Roshambo.css';

import Player from './Player';
import Scoreboard from './Scoreboard';
import Computer from './Computer';

class Roshambo extends Component {
  constructor() {
    super();

    this.state = {
      score: {
        draw: 0,
        win: 0,
        loss: 0
      },
      history: []
    };
  }

  resetGame = () => {
    this.setState({
      score: {
        draw: 0,
        win: 0,
        loss: 0
      },
      history: []
    });
  }

  playerChoice = shape => {
    const { history } = this.state;

    // add the player's choice of shape to the `history` state
    history.push({player: shape});

    this.setState({
      history
    });
  }

  computerChoice = seed => {
    // generate random number between 1 and 3 (repeatable with `seed` for testing)
    return rnd.random( seed || rnd.genCsSeed(), 1, 3 ).value;
  }

  render() {
    return (
      <div className="Roshambo">
        <header className="Roshambo-header">
          <img src={logo} className="Roshambo-logo" alt="logo" />
          <h1 className="Roshambo-title">Welcome to Roshambo</h1>
        </header>
        <p className="Roshambo-intro">
          To get started, edit <code>src/Roshambo.js</code> and save to reload.
        </p>
        <div>
          <Player
            playerChoice={this.playerChoice}
          />
          <Scoreboard
            score={this.state.score}
            resetGame={this.resetGame}
          />
          <Computer />
        </div>
      </div>
    );
  }
}

export default Roshambo;
