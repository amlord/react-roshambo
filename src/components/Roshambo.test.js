import React from 'react';
import ReactDOM from 'react-dom';
import { shallow } from 'enzyme';

import Roshambo from './Roshambo';
import { ROCK, PAPER, SCISSORS, WIN, DRAW, LOSS } from '../helpers/constants';

describe('Roshambo', () => {
  const roshambo = shallow(<Roshambo />);

  it('renders correctly', () => {
    expect(roshambo).toMatchSnapshot();
  });

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Roshambo />, div);
  });

  it('initialises `state` with zeros for `draw`, `loss` and `win`', () => {
    expect(roshambo.state().score).toEqual({
      draw: 0,
      loss: 0,
      win: 0
    });
  });

  it('initialises `state` with an empty history', () => {
    expect(roshambo.state().history).toEqual([]);
  });

  describe('`Player` resets the game', () => {
    beforeEach( () => {
      roshambo.setState({
        score: {
          draw: 1,
          loss: 1,
          win: 1
        },
        history: [
          {player: 1, computer: 1},
          {player: 1, computer: 2},
          {player: 2, computer: 1}
        ]
      });

      roshambo.instance().resetGame();
    });

    it('scores reset to 0', () => {
      expect(roshambo.state().score).toEqual({
        draw: 0,
        loss: 0,
        win: 0
      });
    });

    it('history array emptied', () => {
      expect(roshambo.state().history).toEqual([]);
    });
  });

  describe('`Player` chooses a shape', () => {
    beforeEach( () => {
      roshambo.instance().resetGame();
    });

    it('rock', () => {
      roshambo.instance().playerChoice(ROCK);
      expect(roshambo.state().history[0].player).toEqual(ROCK);
    });

    it('paper', () => {
      roshambo.instance().playerChoice(PAPER);
      expect(roshambo.state().history[0].player).toEqual(PAPER);
    });

    it('scissors', () => {
      roshambo.instance().playerChoice(SCISSORS);
      expect(roshambo.state().history[0].player).toEqual(SCISSORS);
    });
  });

  describe('`Computer` chooses a shape', () => {
    let seed;

    // outcomes within required range
    it('rock', () => {
      // loop through 30 randomly generated outcomes
      for(let i = 0; i < 30; i++) {
        let random = roshambo.instance().computerChoice();
        expect(random).toBeGreaterThanOrEqual(1)
        expect(random).toBeLessThanOrEqual(3);
      }
    });

    // ROCK outcome
    it('rock', () => {
      // `seed` causes outcome of 1 (ROCK)
      seed = [3225582889, 3161022486, 617958775, 3943669506];
      expect(roshambo.instance().computerChoice(seed)).toEqual(ROCK);
    });

    // PAPER outcome
    it('paper', () => {
      // `seed` causes outcome of 2 (PAPER)
      seed = [1421904966, 3555913741, 725236340, 399219481];
      expect(roshambo.instance().computerChoice(seed)).toEqual(PAPER);
    });

    // SCISSORS outcome
    it('scissors', () => {
      // `seed` causes outcome of 3 (SCISSORS)
      seed = [2390802463, 2352622131, 965876151, 1929502849];
      expect(roshambo.instance().computerChoice(seed)).toEqual(SCISSORS);
    });
  });

  describe('game outcome', () => {
    // rock combinations
    it('rock / rock', () => {
      expect(roshambo.instance().gameOutcome(ROCK, ROCK)).toEqual(DRAW);
    });
    it('rock / paper', () => {
      expect(roshambo.instance().gameOutcome(ROCK, PAPER)).toEqual(LOSS);
    });
    it('rock / scissors', () => {
      expect(roshambo.instance().gameOutcome(ROCK, SCISSORS)).toEqual(WIN);
    });

    // paper combinations
    it('paper / rock', () => {
      expect(roshambo.instance().gameOutcome(PAPER, ROCK)).toEqual(WIN);
    });
    it('paper / paper', () => {
      expect(roshambo.instance().gameOutcome(PAPER, PAPER)).toEqual(DRAW);
    });
    it('paper / scissors', () => {
      expect(roshambo.instance().gameOutcome(PAPER, SCISSORS)).toEqual(LOSS);
    });

    // scissors combinations
    it('scissors / rock', () => {
      expect(roshambo.instance().gameOutcome(SCISSORS, ROCK)).toEqual(LOSS);
    });
    it('scissors / paper', () => {
      expect(roshambo.instance().gameOutcome(SCISSORS, PAPER)).toEqual(WIN);
    });
    it('scissors / scissors', () => {
      expect(roshambo.instance().gameOutcome(SCISSORS, SCISSORS)).toEqual(DRAW);
    });
  });

  describe('`history` array updated with computer choice and outcome', () => {
    beforeAll( () => {
      roshambo.instance().resetGame();
    });

    it('item added', () => {
    roshambo.instance().playerChoice(ROCK);
      expect(roshambo.state().history.length).toEqual(1);
    });

    it('contains correct properties', () => {
    expect(roshambo.state().history[0]).toHaveProperty('player');
      expect(roshambo.state().history[0]).toHaveProperty('computer');
      expect(roshambo.state().history[0]).toHaveProperty('outcome');
    });
  });
});