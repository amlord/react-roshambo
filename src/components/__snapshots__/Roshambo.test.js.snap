// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Roshambo renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Roshambo />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <header
          className="Roshambo-header"
>
          <img
                    alt="logo"
                    className="Roshambo-logo"
                    src="logo.svg"
          />
          <h1
                    className="Roshambo-title"
          >
                    Welcome to Roshambo
          </h1>
</header>,
        <p
          className="Roshambo-intro"
>
          To get started, edit 
          <code>
                    src/Roshambo.js
          </code>
           and save to reload.
</p>,
        <div>
          <Player
                    playerChoice={[Function]}
          />
          <Scoreboard
                    resetGame={[Function]}
                    score={
                              Object {
                                        "draw": 0,
                                        "loss": 0,
                                        "win": 0,
                                      }
                    }
          />
          <Computer />
</div>,
      ],
      "className": "Roshambo",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <img
              alt="logo"
              className="Roshambo-logo"
              src="logo.svg"
/>,
            <h1
              className="Roshambo-title"
>
              Welcome to Roshambo
</h1>,
          ],
          "className": "Roshambo-header",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "alt": "logo",
              "className": "Roshambo-logo",
              "src": "logo.svg",
            },
            "ref": null,
            "rendered": null,
            "type": "img",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Welcome to Roshambo",
              "className": "Roshambo-title",
            },
            "ref": null,
            "rendered": "Welcome to Roshambo",
            "type": "h1",
          },
        ],
        "type": "header",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "To get started, edit ",
            <code>
              src/Roshambo.js
</code>,
            " and save to reload.",
          ],
          "className": "Roshambo-intro",
        },
        "ref": null,
        "rendered": Array [
          "To get started, edit ",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "src/Roshambo.js",
            },
            "ref": null,
            "rendered": "src/Roshambo.js",
            "type": "code",
          },
          " and save to reload.",
        ],
        "type": "p",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <Player
              playerChoice={[Function]}
/>,
            <Scoreboard
              resetGame={[Function]}
              score={
                            Object {
                                          "draw": 0,
                                          "loss": 0,
                                          "win": 0,
                                        }
              }
/>,
            <Computer />,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "playerChoice": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "resetGame": [Function],
              "score": Object {
                "draw": 0,
                "loss": 0,
                "win": 0,
              },
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <header
            className="Roshambo-header"
>
            <img
                        alt="logo"
                        className="Roshambo-logo"
                        src="logo.svg"
            />
            <h1
                        className="Roshambo-title"
            >
                        Welcome to Roshambo
            </h1>
</header>,
          <p
            className="Roshambo-intro"
>
            To get started, edit 
            <code>
                        src/Roshambo.js
            </code>
             and save to reload.
</p>,
          <div>
            <Player
                        playerChoice={[Function]}
            />
            <Scoreboard
                        resetGame={[Function]}
                        score={
                                    Object {
                                                "draw": 0,
                                                "loss": 0,
                                                "win": 0,
                                              }
                        }
            />
            <Computer />
</div>,
        ],
        "className": "Roshambo",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <img
                alt="logo"
                className="Roshambo-logo"
                src="logo.svg"
/>,
              <h1
                className="Roshambo-title"
>
                Welcome to Roshambo
</h1>,
            ],
            "className": "Roshambo-header",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "alt": "logo",
                "className": "Roshambo-logo",
                "src": "logo.svg",
              },
              "ref": null,
              "rendered": null,
              "type": "img",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Welcome to Roshambo",
                "className": "Roshambo-title",
              },
              "ref": null,
              "rendered": "Welcome to Roshambo",
              "type": "h1",
            },
          ],
          "type": "header",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "To get started, edit ",
              <code>
                src/Roshambo.js
</code>,
              " and save to reload.",
            ],
            "className": "Roshambo-intro",
          },
          "ref": null,
          "rendered": Array [
            "To get started, edit ",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "src/Roshambo.js",
              },
              "ref": null,
              "rendered": "src/Roshambo.js",
              "type": "code",
            },
            " and save to reload.",
          ],
          "type": "p",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <Player
                playerChoice={[Function]}
/>,
              <Scoreboard
                resetGame={[Function]}
                score={
                                Object {
                                                "draw": 0,
                                                "loss": 0,
                                                "win": 0,
                                              }
                }
/>,
              <Computer />,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "playerChoice": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "resetGame": [Function],
                "score": Object {
                  "draw": 0,
                  "loss": 0,
                  "win": 0,
                },
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
